fabric-resource-loader-v1/src/main/java/net/fabricmc/fabric/impl/resource/v1/ResourceLoaderImpl.java

```diff
@@ -0,0 +1,276 @@
+import net.fabricmc.fabric.api.util.TriState;
+import net.fabricmc.fabric.impl.base.toposort.NodeSorting;
+import net.fabricmc.fabric.impl.base.toposort.SortableNode;
+import net.fabricmc.loader.api.FabricLoader;
+
+public final class ResourceLoaderImpl implements ResourceLoader {
+	private static final Logger LOGGER = LoggerFactory.getLogger("ResourceLoader");
+	private static final Map<ResourceType, ResourceLoaderImpl> IMPL_MAP = new EnumMap<>(ResourceType.class);
+
+	private static final boolean DEBUG_RELOADERS_IDENTITY = TriState.fromSystemProperty("fabric.resource_loader.debug.reloaders_identity")
+			.orElse(FabricLoader.getInstance().isDevelopmentEnvironment());
+	private static final boolean DEBUG_RELOADERS_ORDER = TriState.fromSystemProperty("fabric.resource_loader.debug.reloaders_order")
+			.orElse(true);
+
+	public static ResourceLoaderImpl get(ResourceType type) {
+		return IMPL_MAP.computeIfAbsent(type, ResourceLoaderImpl::new);
+	}
+
+	private final Set<Identifier> addedReloaderIds = new ObjectOpenHashSet<>();
+	private final Set<IdentifiableResourceReloader> addedReloaders = new LinkedHashSet<>();
+	private final Set<ReloaderOrder> reloadersOrdering = new LinkedHashSet<>();
+	private final ResourceType type;
+
+	private ResourceLoaderImpl(ResourceType type) {
+		this.type = type;
+	}
+
+	@Override
+	public void registerReloader(IdentifiableResourceReloader reloader) {
+		if (!this.addedReloaderIds.add(reloader.getFabricId())) {
+			throw new IllegalStateException(
+					"Tried to register resource reloader %s twice!".formatted(reloader.getFabricId())
+			);
+		}
+
+		if (!this.addedReloaders.add(reloader)) {
+			throw new IllegalStateException(
+					"Resource reloader with previously unknown ID %s already in resource reloader set!"
+							.formatted(reloader.getFabricId())
+			);
+		}
+	}
+
+	@Override
+	public void addReloaderOrdering(@NotNull Identifier firstReloader, @NotNull Identifier secondReloader) {
+		Objects.requireNonNull(firstReloader, "The first reloader identifier should not be null.");
+		Objects.requireNonNull(secondReloader, "The second reloader identifier should not be null.");
+
+		if (firstReloader.equals(secondReloader)) {
+			throw new IllegalArgumentException("Tried to add a phase that depends on itself.");
+		}
+
+		this.reloadersOrdering.add(new ReloaderOrder(firstReloader, secondReloader));
+	}
+
+	public static List<ResourceReloader> sort(ResourceType type, List<ResourceReloader> listeners) {
+		if (type == null) {
+			return listeners;
+		}
+
+		ResourceLoaderImpl instance = get(type);
+
+		if (instance != null) {
+			var mutable = new ArrayList<>(listeners);
+			instance.sort(mutable);
+			return Collections.unmodifiableList(mutable);
+		}
+
+		return listeners;
+	}
+
+	/**
+	 * Sorts the given resource reloaders to satisfy dependencies.
+	 *
+	 * @param reloaders the resource reloaders to sort
+	 */
+	private void sort(List<ResourceReloader> reloaders) {
```

Comment:
Have you thought about writing some unit tests for this? There is a lot of logic with this sorting that is imporant to make sure works as expected. Testmods make it hard to ensure its working as expected especially after future updates. In my particle renderer test I added some unit tests: https://github.com/FabricMC/fabric/pull/4854/files#diff-4b87f72c85dccd53541717078398c23c03b72c10bcf90f1468bf4944dadd0df0R36

Reply:
Great, thanks. Im mostly worried for when someone needs to touch this in the future.